cpp/sophus/geometry/ray.h

```diff
@@ -18,71 +18,93 @@
 
 namespace sophus {
 
+// Forward declarations
+template <class TScalar, int kN>
+class UnitVector;
+template <class TScalar, int kN>
+class Ray;
+
+// Convenience typedefs
+template <class TScalar>
+using UnitVector3 = UnitVector<TScalar, 3>;
+template <class TScalar>
+using UnitVector2 = UnitVector<TScalar, 2>;
+
+using UnitVector2F64 = UnitVector2<double>;
+using UnitVector3F64 = UnitVector3<double>;
+
+template <class TScalar>
+using Ray2 = Ray<TScalar, 2>;
 template <class TScalar>
-class UnitVector3 {
+using Ray3 = Ray<TScalar, 3>;
+
+using Ray2F64 = Ray2<double>;
+using Ray3F64 = Ray3<double>;
+
+template <class TScalar, int kN>
+class UnitVector {
  public:
   // Precondition: v must be of unit length.
-  static UnitVector3 fromUnitVector(Eigen::Matrix<TScalar, 3, 1> const& v) {
-    farm_ng::Expected<UnitVector3> e_vec = tryFromUnitVector(v);
+  static UnitVector fromUnitVector(Eigen::Matrix<TScalar, kN, 1> const& v) {
+    farm_ng::Expected<UnitVector> e_vec = tryFromUnitVector(v);
     if (!e_vec.has_value()) {
       FARM_FATAL("{}", e_vec.error());
     }
     return e_vec.value();
   }
 
-  static farm_ng::Expected<UnitVector3> tryFromUnitVector(
-      Eigen::Matrix<TScalar, 3, 1> const& v) {
+  static farm_ng::Expected<UnitVector> tryFromUnitVector(
+      Eigen::Matrix<TScalar, kN, 1> const& v) {
     using std::abs;
-    if ((v.squaredNorm() - TScalar(1.0)) < sophus::kEpsilon<TScalar>) {
```

Comment:
@strasdat This is the location of the bug and it's worth checking I got it right!